### PR TITLE
syslinux: add uefi support

### LINKED ISSUE #41403: pxelinux doesn't support UEFI
## Issue description

pxelinux (from syslinux) is compiled without support for UEFI PXE boot: https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/syslinux/default.nix#L47

I had a quick look at getting it to compile uefi support but I didn't get very far.

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.16.11, NixOS, 18.09.git.c2a4a5143f2M (Jellyfish)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.0.2`
 - channels(root): `""`
 - channels(peter): `""`
 - nixpkgs: `/home/peter/src/active/nixpkgs_unstable`

### PR DESCRIPTION
**Pull request by ajs124**
###### Motivation for this change
Fixes #41403

The main changes are fetching submodules, using a patch with the same name, but a wider scope, from Arch Linux instead of Debian, dropping some make flags that break the build (yeah…) and a stubs-32.h hack I stole from... somewhere. Xen maybe?

The second commit adds an unofficial and not particularly polished patchset that adds support for EFI binary "chainloading". This successfully executes an UEFI Shell, but fails to run some other things like rEFInd.

It's a separate commit because, while I'd like to see the general UEFI support merged, I'm definitely willing to drop this patcheset, if that's what the maintainers/community prefers.

cc @samueldr @peterhoeg 

###### Things done
Tested booting on UEFI and BIOS on real hardware and in qemu.

### CODE REVIEW

**Comment by peterhoeg** [comment]
The uefi support bit is great, thanks!

I'd be very hesitant to support merging an out of band patch for something as integral as syslinux unless "blessed" by other distributions.

**Comment by ajs124** [comment]
Understandable. I dropped that commit, so now this only adds UEFI support, as it works on upstream syslinux.

This means you can boot into syslinux from UEFI and start a Linux kernel with it, but cannot execute any other UEFI binaries (like an EFI Shell) from it.

File: pkgs/os-specific/linux/syslinux/default.nix [hunk 1]
```diff
@@ -1,14 +1,15 @@
-{ stdenv, fetchFromRepoOrCz, fetchurl, nasm, perl, python3, libuuid, mtools, makeWrapper }:
+{ stdenv, fetchgit, fetchurl, fetchpatch, nasm, perl, python3, libuuid, mtools, makeWrapper }:
 
 stdenv.mkDerivation {
   name = "syslinux-2019-02-07";
```
Comment (peterhoeg) [inline pkgs/os-specific/linux/syslinux/default.nix:4]: Please split this into pname and version.

File: pkgs/os-specific/linux/syslinux/default.nix [hunk 2]
```diff
@@ -1,14 +1,15 @@
...
   # This is syslinux-6.04-pre3^1; syslinux-6.04-pre3 fails to run.
   # Same issue here https://www.syslinux.org/archives/2019-February/026330.html
-  src = fetchFromRepoOrCz {
-    repo = "syslinux";
+  src = fetchgit {
+    url = "https://repo.or.cz/syslinux";
     rev = "b40487005223a78c3bb4c300ef6c436b3f6ec1f7";
-    sha256 = "1qrxl1114sr2i2791z9rf8v53g200aq30f08808d7i8qnmgvxl2w";
+    sha256 = "1acf6byx7i6vz8hq6mra526g8mf7fmfhid211y8nq0v6px7d3aqs";
+    fetchSubmodules = true;
```
Comment (peterhoeg) [inline pkgs/os-specific/linux/syslinux/default.nix:12]: You're dropping `fetchFromRepoOrCz` because of a lack of submodule support?

Comment (ajs124) [reply pkgs/os-specific/linux/syslinux/default.nix:12]: Exactly, yes. gnu-efi is a [submodule](https://repo.or.cz/syslinux.git/blob_plain/HEAD:/.gitmodules).

File: pkgs/os-specific/linux/syslinux/default.nix (hunk 1 above)
Comment (ajs124) [reply pkgs/os-specific/linux/syslinux/default.nix]: done

File: pkgs/os-specific/linux/syslinux/default.nix [hunk 3]
```diff
@@ -1,14 +1,16 @@
-{ stdenv, fetchFromRepoOrCz, fetchurl, nasm, perl, python3, libuuid, mtools, makeWrapper }:
+{ stdenv, fetchgit, fetchurl, fetchpatch, nasm, perl, python3, libuuid, mtools, makeWrapper }:
 
 stdenv.mkDerivation {
-  name = "syslinux-2019-02-07";
+  pname = "syslinux";
+  version = "2019-02-07";
```
Comment (peterhoeg) [inline pkgs/os-specific/linux/syslinux/default.nix:5]: Sorry, last thing, the version should be `version = "unstable-20190207";`

Comment (ajs124) [reply pkgs/os-specific/linux/syslinux/default.nix:5]: Should I maybe put these changes in a separate commit?

I agree that we might want to make these changes, while we're at it, but they seem pretty unrelated to the UEFI stuff.

**Comment by peterhoeg** [comment]
>   Should I maybe put these changes in a separate commit?

We normally do these minor cleanups when the package is being touched anyway.

**Comment by peterhoeg** [comment]
Thanks for all your hard work @ajs124

File: pkgs/os-specific/linux/syslinux/default.nix [hunk 4]
```diff
@@ -20,9 +22,9 @@ stdenv.mkDerivation {
       url = mkURL "fa1349f1" "0002-gfxboot-menu-label.patch";
       sha256 = "06ifgzbpjj4picpj17zgprsfi501zf4pp85qjjgn29i5rs291zni";
     })
-    (fetchurl {
-      url = mkURL "477e56d2" "0005-gnu-efi-version-compatibility.patch";
-      sha256 = "041568b4abb79wynyps1n04lg4fr26rc3sbjncz99pp0mbz0ajlm";
+    (fetchpatch {
+      url = "https://git.archlinux.org/svntogit/packages.git/plain/trunk/0005-gnu-efi-version-compatibility.patch?h=packages/syslinux";
```
Comment (jtojnar) [inline pkgs/os-specific/linux/syslinux/default.nix:26]: This is not a static URL, you need to pin it to Arch git commit.

File: pkgs/os-specific/linux/syslinux/default.nix [hunk 5]
```diff
@@ -48,6 +50,10 @@ stdenv.mkDerivation {
     # fix tests
     substituteInPlace tests/unittest/include/unittest/unittest.h \
       --replace /usr/include/ ""
+
+    # Hack to get `gcc -m32' to work without having 32-bit Glibc headers.
+    mkdir gnu-efi/inc/ia32/gnu
```
Comment (jtojnar) [inline pkgs/os-specific/linux/syslinux/default.nix:55]: Is it possible to use gnu-efi from the nixpkgs package instead of a submodule?

**Pull request merged**